### PR TITLE
ENT-4125 tabular display of catalog

### PR DESCRIPTION
**Pull request by binodpant**
JIRA: https://openedx.atlassian.net/browse/ENT-4125
Introduces data table into the enterpriseCatalogs component, runs a query with no filters against the catalog index

Tested from staging algolia

things to fix before Merge:

- [x] Need paragon update once it releases from ENT-4290
- [X] styling changes part of this PR or no, did requisite styles

looks like this right now

Screen Shot 2021-03-10 at 4 19 25 PM

<hr/>

Screen Shot 2021-03-10 at 4 19 18 PM

Screen Shot 2021-03-10 at 4 21 39 PM

Screen Shot 2021-03-10 at 4 23 53 PM

### CODE REVIEW

File: src/components/LoadingMessage/index.jsx [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+const LoadingMessage = (props) => {
+  const { className } = props;
+  return (
+    <div
+      className={classNames(
+        'loading d-flex',
+        className,
+      )}
+    >
+      Loading...
```
Comment (raq929) [inline src/components/LoadingMessage/index.jsx]: This loading module is *terrible*, let's not re-use it. In other places in learner-portal we're using `react-skeleton`, it'll create a much nicer loading experience.

File: src/components/NotFoundPage.jsx [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import React from 'react';
+import { Container } from '@edx/paragon';
+import { Helmet } from 'react-helmet';
+import { PAGE_TITLE } from '../constants';
+
+const NotFoundPage = () => (
+  <Container size="lg" className="mt-3">
+    <Helmet title={PAGE_TITLE} />
+    <div className="text-center py-5">
+      This content is not available
```
Comment (raq929) [inline src/components/NotFoundPage.jsx]: Check with UX to be sure this is the text we want.

File: src/components/NotFoundPage.jsx [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import React from 'react';
+import { Container } from '@edx/paragon';
+import { Helmet } from 'react-helmet';
+import { PAGE_TITLE } from '../constants';
+
+const NotFoundPage = () => (
+  <Container size="lg" className="mt-3">
+    <Helmet title={PAGE_TITLE} />
```
Comment (raq929) [inline src/components/NotFoundPage.jsx]: This should be the `PageWrapper` component.

File: src/components/PageWrapper.jsx [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import { Container } from '@edx/paragon';
+import { Helmet } from 'react-helmet';
+
+import { PAGE_TITLE } from '../constants';
+
+// eslint-disable-next-line react/prop-types
+const Wrapper = ({ children }) => (
+  <Container size="lg" className="mt-3">
+    <Helmet title={PAGE_TITLE} />
+    <div className="text-center py-5">
```
Comment (raq929) [inline src/components/PageWrapper.jsx]: I think it's unlikely that we'll want `text-center` on an entire page?

File: src/index.jsx [hunk 5]
```diff
@@ -30,6 +28,15 @@ subscribe(APP_INIT_ERROR, (error) => {
 });
 
 initialize({
+  handlers: {
+    config: () => {
+      mergeConfig({
```
Comment (raq929) [inline src/index.jsx:33]: Yes, awesome, this is definitely the way to be doing this. 
A part of this ticket should also be making a pr on `edx-internal` to add the algolia variables there.

File: src/store.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import { applyMiddleware, createStore } from 'redux';
```
Comment (raq929) [inline src/store.js]: Let's *not* add redux into this project. We don't need it. We can use the `useReducer` hook if we need a reducer, and we can use React context if we need a store.

**Review by raq929 — APPROVED** [review]

File: src/store.js (hunk 6 above)
Comment (binodpant) [reply src/store.js]: yeah agreed I had started with a store and now I don't want it

File: src/index.jsx (hunk 5 above)
Comment (binodpant) [reply src/index.jsx:33]: filed: https://openedx.atlassian.net/browse/ENT-4289 for that

Comment (binodpant) [reply src/index.jsx:33]: will take care of it before merging this one, it's very simple and small

File: src/store.js (hunk 6 above)
Comment (binodpant) [reply src/store.js]: I remove the store which is not used anyway, I could not remove `react-redux` itself as a dep due to peer dep but yeah it does not use redux store directly anymore

File: src/components/PageWrapper.jsx (hunk 4 above)
Comment (binodpant) [reply src/components/PageWrapper.jsx]: yes removed this, don't need on entire page and in fact looks better without it

File: src/components/NotFoundPage.jsx (hunk 2 above)
Comment (binodpant) [reply src/components/NotFoundPage.jsx]: ran by Joe, updated text

File: src/index.jsx (hunk 5 above)
Comment (binodpant) [reply src/index.jsx:33]: actually no need to hold up this pr for the ENT-4289 work, that will happen once Talia deploys MFE, and merging this in won't deploy it anywhere yet anyway

File: .gitignore [hunk 7]
```diff
@@ -1,7 +1,7 @@
 # local dev related
 # https://github.com/edx/frontend-build#local-module-configuration-for-webpack
 .env.private
-module.config.js
+module.config.js*
```
Comment (raq929) [inline .gitignore:4]: Why the * ?

File: package.json [hunk 8]
```diff
@@ -37,21 +37,29 @@
...
+    "algoliasearch": "4.8.5",
     "babel-polyfill": "6.26.0",
+    "classnames": "^2.2.6",
     "prop-types": "15.7.2",
     "react": "16.12.0",
     "react-dom": "16.12.0",
-    "react-redux": "7.1.3",
+    "react-helmet": "6.1.0",
+    "react-instantsearch-dom": "6.8.3",
+    "react-loading-skeleton": "^2.2.0",
```
Comment (raq929) [inline package.json]: This will likely get autofixed if we have the dependency bot, but we shouldn't have the `^` anymore.

File: src/components/PageWrapper.jsx [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import { Container } from '@edx/paragon';
+import { Helmet } from 'react-helmet';
+
+import { PAGE_TITLE } from '../constants';
+
+// eslint-disable-next-line react/prop-types
+const Wrapper = ({ children }) => (
+  <Container size="lg" className="mt-3">
```
Comment (raq929) [inline src/components/PageWrapper.jsx]: What happens on a small screen?

File: src/components/catalogs/CatalogSearchResults.jsx [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+const CatalogSearchResults = ({
+  searchResults,
+  // algolia recommends this prop instead of searching
+  isSearchStalled,
+  searchState,
+  error,
+}) => {
+  if (isSearchStalled) {
+    return (<Skeleton height={25} />);
```
Comment (raq929) [inline src/components/catalogs/CatalogSearchResults.jsx]: I think we probably want more than this? Like a skeleton of a table?
Can you put up a screen shot of what this looks like currently?

File: src/components/catalogs/CatalogSearchResults.jsx [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+  }
+  if (error) {
+    return (
+      <StatusAlert
```
Comment (raq929) [inline src/components/catalogs/CatalogSearchResults.jsx]: I think there's a warning about the StatusAlert in the tests in AdminPortal. I think if we're adding it here, we should fix the warnings.

File: src/components/catalogs/CatalogSearchResults.jsx [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+      Header: TABLE_HEADERS.subject,
+      accessor: 'subjects[0]',
+    },
+    {
+      Header: TABLE_HEADERS.partner,
+      accessor: 'partners[0].name',
+    },
+  ], []);
+
+  const calculatePage = () => {
```
Comment (raq929) [inline src/components/catalogs/CatalogSearchResults.jsx]: If you are defining it this function here inside the component, the `useMemo` below is kind of useless. This will always register as a new function and the code inside the `useMemo` will always run.
I'd suggest either moving the definition outside the component, which I like because that makes it testable, or just not memoizing, because this isn't a high-cost operation, so memoizing doesn't seem like a necessary optimization.

File: src/components/catalogs/CatalogSearchResults.jsx [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+  const tableData = useMemo(() => searchResults?.hits || [], [searchResults?.hits]);
+  return (
+    <>
+      <div>
+        <DataTable
+          columns={columns}
+          data={tableData}
+          itemCount={searchResults?.nbHits}
+          isSelectable
```
Comment (raq929) [inline src/components/catalogs/CatalogSearchResults.jsx]: Does this table actually need to be selectable? Are people going to perform bulk actions on it? (I don't remember the spec.)

File: src/components/catalogs/EnterpriseCatalogs.jsx [hunk 14]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+import {
+  Configure, InstantSearch,
+} from 'react-instantsearch-dom';
+
+import { SearchData, SearchHeader } from '@edx/frontend-enterprise';
+
+import Wrapper from '../PageWrapper';
```
Comment (raq929) [inline src/components/catalogs/EnterpriseCatalogs.jsx]: Let's call it `PageWrapper`; imports get really confusing if we rename them; people used to React will expect this to have the same name as it's file name.

File: src/components/catalogs/EnterpriseCatalogs.jsx [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+  return (
+    <Wrapper>
+      <SearchData>
+        <InstantSearch
+          indexName={algoliaIndexName}
+          searchClient={searchClient}
+        >
+          <Configure hitsPerPage={NUM_RESULTS_PER_PAGE} />
+          <div className="search-header-wrapper">
```
Comment (raq929) [inline src/components/catalogs/EnterpriseCatalogs.jsx]: I expect the SearchHeader doesn't currently accept a `className` prop, but for future reference, I'd prefer to pass this in as a prop rather than adding an extra div.

Comment (raq929) [reply src/components/catalogs/EnterpriseCatalogs.jsx]: Also, I'm not seeing this class used?

File: src/index.scss [hunk 16]
```diff
@@ -5,3 +5,10 @@
 
 @import "~@edx/frontend-component-header/dist/index";
 @import "~@edx/frontend-component-footer/dist/footer";
+
+@import "~@edx/frontend-enterprise/src/course-search/styles/index";
+
+.bg-brand-primary {
```
Comment (raq929) [inline src/index.scss]: Please wrap this in your search header class. `bg-brand-primary` is a css class that might be re-used. We don't want this to accidentally override other styles.
Also, please move this css into the `components/catalogs` folder and import it here. I know we do this in admin portal, but in general, css should live with it's components and the `index.scss` file should *only* contain imports.

**Review by raq929 — CHANGES_REQUESTED** [review]
I've got a few review comments, and this needs tests before it can be merged! Looking good, though!

File: .gitignore (hunk 7 above)
Comment (binodpant) [reply .gitignore:4]: ah yeah I found it convenient to often rename my module.config.js to module.config.js.unused or similar rather than delete and forget what I need.. so this helps with that while keeping current behavior

File: package.json (hunk 8 above)
Comment (binodpant) [reply package.json]: yeah I missed this one but will clean it up here instead of waiting for the tool ty

File: src/components/PageWrapper.jsx (hunk 9 above)
Comment (binodpant) [reply src/components/PageWrapper.jsx]: it works just fine, and checked the doc, 
```
<Container size="lg" className="bg-danger-300 mb-4">
  The content in this container won't exceed the large width.
</Container>
```

 it appears the lg is mainly about not extending beyond a certain width

File: src/components/catalogs/CatalogSearchResults.jsx (hunk 10 above)
Comment (binodpant) [reply src/components/catalogs/CatalogSearchResults.jsx]: hmm when I forced it to show always to see what this shows, and right now I. just see this (should I add a paragon loading spinner in side of this here?)

Screen Shot 2021-03-10 at 11 36 44 AM

Comment (binodpant) [reply src/components/catalogs/CatalogSearchResults.jsx]: it's just a animated hardly visible gray bar right now

Comment (binodpant) [reply src/components/catalogs/CatalogSearchResults.jsx]: ah single line skeleton right now and small height so looks hard to find, what about this?
Screen Shot 2021-03-10 at 11 45 39 AM

File: src/components/catalogs/CatalogSearchResults.jsx (hunk 11 above)
Comment (binodpant) [reply src/components/catalogs/CatalogSearchResults.jsx]: yeah I just saw onClose is a required field somehow.. .perhaps we need to add it even if not used?

Comment (binodpant) [reply src/components/catalogs/CatalogSearchResults.jsx]: hmm also per paragon docs, looks like Alert should be now used instead of StatusAlert anyway, will make that change

Comment (binodpant) [reply src/components/catalogs/CatalogSearchResults.jsx]: now looks like this 
Screen Shot 2021-03-10 at 12 11 04 PM

File: src/components/catalogs/CatalogSearchResults.jsx (hunk 12 above)
Comment (binodpant) [reply src/components/catalogs/CatalogSearchResults.jsx]: yeah good callout this is leftover from my attempts to debug issue, don't need this useMemo here

Comment (binodpant) [reply src/components/catalogs/CatalogSearchResults.jsx]: think this is simpler 

```
  const page = refinementsFromQueryParams.page || (searchState ? searchState.page : 0);
```

File: src/components/catalogs/CatalogSearchResults.jsx (hunk 13 above)
Comment (binodpant) [reply src/components/catalogs/CatalogSearchResults.jsx]: nope no use case right now for selectable

File: src/components/catalogs/EnterpriseCatalogs.jsx (hunk 15 above)
Comment (binodpant) [reply src/components/catalogs/EnterpriseCatalogs.jsx]: ah yes removing this since this class is not even defined, and it has no visual effect

File: src/index.scss (hunk 16 above)
Comment (binodpant) [reply src/index.scss]: good suggestion, done!

**Comment by binodpant** [comment]
all feedback applied!

File: src/components/catalogs/styles/_index.scss [hunk 17]
```diff
@@ -0,0 +1,13 @@
+.enterprise-catalogs-header {
```
Comment (raq929) [inline src/components/catalogs/styles/_index.scss:1]: 🎉

File: src/index.jsx [hunk 18]
```diff
@@ -1,26 +1,24 @@
...
+
+import App from './components/app/App';
 
 import appMessages from './i18n';
 
 import './index.scss';
 
 subscribe(APP_READY, () => {
   ReactDOM.render(
-    <AppProvider>
```
Comment (raq929) [inline src/index.jsx:19]: Are we not using AppProvider because we don't need authentication?

File: src/components/catalogs/CatalogSearchResults.jsx (hunk 10 above)
Comment (raq929) [reply src/components/catalogs/CatalogSearchResults.jsx]: Yeah, that'll do for the moment. 😄  We can run it by UX later.

File: src/components/catalogs/CatalogSearchResults.jsx (hunk 11 above)
Comment (raq929) [reply src/components/catalogs/CatalogSearchResults.jsx]: Throw a top margin on that, and I think we're good.

File: src/index.jsx (hunk 18 above)
Comment (binodpant) [reply src/index.jsx:19]: oh wait ok I do  use it, just in the <App> component

File: src/components/catalogs/CatalogSearchResults.jsx [hunk 19]
```diff
@@ -10,21 +10,34 @@ import {
...
 };
 
-const CatalogSearchResults = ({
+/**
+ * The core search resultss rendering component.
+ *
+ * Wrapping this in `connectStateResults()` will inject the first few props.
+ *
+ * @param {object} args arguments
+ * @param {object} args.searchResults Results of search (see: `connectStateResults``)
```
Comment (binodpant) [inline src/components/catalogs/CatalogSearchResults.jsx:29]: I do not like the 'hidden-ness' of props like this that come from connected/HOC components, so I feel better when I doc it like this ...

File: src/components/catalogs/CatalogSearchResults.jsx (hunk 11 above)
Comment (binodpant) [reply src/components/catalogs/CatalogSearchResults.jsx]: done! mt-2 for both alerts

File: src/components/catalogs/CatalogSearchResults.jsx (hunk 19 above)
Comment (raq929) [reply src/components/catalogs/CatalogSearchResults.jsx:29]: This is what PropTypes is for? I don't see the point of further documenting it like this, but up to you.

File: src/components/catalogs/CatalogSearchResults.jsx [hunk 20]
```diff
@@ -10,21 +10,34 @@ import {
...
+ * @param {object} args.error Error with `message` field if available (see: `connectStateResults``)
+ * @param {object} args.paginationComponent Defaults to <SearchPagination> but can be injected
+ */
+export const CoreCatalogSearchResults = ({
   searchResults,
   // algolia recommends this prop instead of searching
   isSearchStalled,
   searchState,
   error,
+  paginationComponent: PaginationComponent = SearchPagination,
```
Comment (raq929) [inline src/components/catalogs/CatalogSearchResults.jsx]: At edx we generally use `defaultProps` rather than doing it like this, but I don't really have an issue with it.

File: src/components/catalogs/CatalogSearchResults.jsx [hunk 21]
```diff
@@ -107,6 +120,7 @@ CatalogSearchResults.propTypes = {
   searchState: PropTypes.shape({
     page: PropTypes.number,
   }).isRequired,
+  paginationComponent: PropTypes.node.isRequired,
 };
 
-export default connectStateResults(CatalogSearchResults);
+export default connectStateResults(CoreCatalogSearchResults);
```
Comment (raq929) [inline src/components/catalogs/CatalogSearchResults.jsx]: Please use `BaseCatalogSearchResults` instead of `Core`, to be conform with standards used in other repos.

File: src/components/NotFoundPage.jsx [hunk 22]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+import PageWrapper from './PageWrapper';
+
+const NotFoundPage = () => (
```
Comment (raq929) [inline src/components/NotFoundPage.jsx:5]: It's a very simple page right now, but let's add a test for this component.

File: src/components/PageWrapper.jsx [hunk 23]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import { Container } from '@edx/paragon';
+import { Helmet } from 'react-helmet';
+
+import { PAGE_TITLE } from '../constants';
+
+// eslint-disable-next-line react/prop-types
+const Wrapper = ({ children }) => (
```
Comment (raq929) [inline src/components/PageWrapper.jsx:10]: This component also needs tests.

File: src/components/app/App.jsx [hunk 24]
```diff
@@ -0,0 +1,22 @@
...
+
+import { EnterpriseCatalogs } from '../catalogs';
+import NotFoundPage from '../NotFoundPage';
+
+export default function App() {
+  return (
+    <AppProvider>
+      <Header />
+      <Switch>
+        <PageRoute exact path="/" component={EnterpriseCatalogs} />
```
Comment (raq929) [inline src/components/app/App.jsx]: I'd like to start this repo off right with testing, and test our routes. 
Example here: https://techdoma.in/react-js-testing/how-to-test-routes-react-router-dom-in-react-using-jest-1

File: src/components/catalogs/CatalogSearchResults.jsx [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+export const CoreCatalogSearchResults = ({
+  searchResults,
+  // algolia recommends this prop instead of searching
+  isSearchStalled,
+  searchState,
+  error,
+  paginationComponent: PaginationComponent = SearchPagination,
+}) => {
+  if (isSearchStalled) {
+    return (<Skeleton className="m-1 loading-skeleton" height={25} count={5} />);
```
Comment (raq929) [inline src/components/catalogs/CatalogSearchResults.jsx]: We should test that this is displayed when the `isSearchStalled` is true.

File: src/components/catalogs/CatalogSearchResults.jsx [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+      <Alert className="mt-2" variant="warning">
+        {NO_DATA_MESSAGE}
+      </Alert>
+    );
+  }
+
+  const { refinementsFromQueryParams } = useContext(SearchContext);
+  const columns = useMemo(() => [
+    {
+      Header: TABLE_HEADERS.courseName,
```
Comment (raq929) [inline src/components/catalogs/CatalogSearchResults.jsx:73]: We should test that the correct table headers are displayed in the document.

File: src/components/catalogs/EnterpriseCatalogs.jsx [hunk 27]
```diff
@@ -0,0 +1,30 @@
...
+} from 'react-instantsearch-dom';
+
+import { SearchData, SearchHeader } from '@edx/frontend-enterprise';
+
+import PageWrapper from '../PageWrapper';
+import { NUM_RESULTS_PER_PAGE } from '../../constants';
+import CatalogSearchResults from './CatalogSearchResults';
+import { useAlgoliaIndex } from './data/hooks';
+
+export default function EnterpriseCatalogs() {
```
Comment (raq929) [inline src/components/catalogs/EnterpriseCatalogs.jsx:13]: I don't think this component needs a lot in terms of testing, but we should test it if at all possible.

File: src/components/catalogs/data/hooks.jsx [hunk 28]
```diff
@@ -0,0 +1,21 @@
+/* eslint-disable import/prefer-default-export */
+import { useMemo } from 'react';
+
+import algoliasearch from 'algoliasearch/lite';
+
+import { getConfig } from '@edx/frontend-platform';
+
+const useAlgoliaIndex = () => {
```
Comment (raq929) [inline src/components/catalogs/data/hooks.jsx:8]: Please test this hook.

**Review by raq929 — COMMENTED** [review]
TL; DR More tests please!
What you've got is looking good!

File: src/components/catalogs/EnterpriseCatalogs.jsx (hunk 27 above)
Comment (raq929) [reply src/components/catalogs/EnterpriseCatalogs.jsx:13]: I couldn't find a great way to mock Algolia, though, so... this is the only one I'm kind of willing to skip if necessary.

File: src/components/catalogs/CatalogSearchResults.jsx (hunk 19 above)
Comment (binodpant) [reply src/components/catalogs/CatalogSearchResults.jsx:29]: yeah some of this may come down to taste / prefs too totally, I just ilke being able to read in the function doc what things are for etc, proptypes tells me the type but not the purpose esp. from 'magical' places like connected components

File: src/components/catalogs/CatalogSearchResults.jsx (hunk 21 above)
Comment (binodpant) [reply src/components/catalogs/CatalogSearchResults.jsx]: this is probably just my years of OOP muscle brain haha. `Base` makes me think of inheritance or components down from there.. and there are not. But I also see consistency as a good way to help understand so happy to change this instead of adding yet another prefix term!

File: src/components/app/App.jsx (hunk 24 above)
Comment (binodpant) [reply src/components/app/App.jsx]: yeah there are more than one route, so testing routes is a good idea

File: src/components/catalogs/EnterpriseCatalogs.jsx (hunk 27 above)
Comment (binodpant) [reply src/components/catalogs/EnterpriseCatalogs.jsx:13]: this one could get a bit tricky but let me see what can be done.. if seems too much work maybe I will let it be, but yeah worth a try

File: src/index.jsx (hunk 18 above)
Comment (raq929) [reply src/index.jsx:19]: If you're using it, it belongs here; it needs to wrap the entire app, not just the component. Things will get *very* confused if it's not at the top level.

File: src/components/catalogs/EnterpriseCatalogs.jsx (hunk 27 above)
Comment (binodpant) [reply src/components/catalogs/EnterpriseCatalogs.jsx:13]: managed to setup a test thatt renders the main component (except the AppProvider part since I want to mock the data that goes into it), using this doc as guide: https://testing-library.com/docs/example-react-router/

File: src/components/app/App.jsx (hunk 24 above)
Comment (binodpant) [reply src/components/app/App.jsx]: tested!

File: src/components/catalogs/CatalogSearchResults.jsx (hunk 26 above)
Comment (binodpant) [reply src/components/catalogs/CatalogSearchResults.jsx:73]: tested!

**Review by raq929 — APPROVED** [review]
I love it! :shipit:
